# PYTHON-12. Продвинутые методы работы с данными в Pandas

## 1. Введение



✍ В этом модуле мы продолжим изучение библиотеки Pandas и рассмотрим более сложные методы работы с данными, которые помогают производить их анализ и подготовку. Модуль можно условно разделить на две части:

в первой части мы поговорим о продвинутых методах Pandas, позволяющих производить более детальный анализ зависимостей в данных и делать на их основе выводы;
вторая часть посвящена работе с несколькими таблицами и их объединению.

Как мы уже говорили, этап подготовки данных играет важную роль в Data Science. Однако этот этап не пройдёт успешно без тщательного анализа зависимостей в данных. Для успешного преобразования таблицы вам необходимо иметь представление о том, как признаки связаны между собой и с целевым признаком, который вы хотели бы предсказать. Поэтому в первой части модуля мы рассмотрим основные инструменты, позволяющие производить такой анализ, а также методы сортировки данных.
Другая проблема заключается в том, что ваши данные не всегда могут быть представлены в виде одной таблицы. Такое нередко случается при работе с выгрузками из целой базы данных. Например, к вам в руки могут попасть несколько таблиц с информацией о клиентах, об их заказах и о поставщиках. Однако вашей целью может быть анализ данных о поставках в целом при использовании всей предоставленной информации. В таком случае таблицы необходимо объединить в единую структуру — об этом мы и поговорим во второй части модуля.

Цели данного модуля:

Научиться сортировать данные.
Получить навыки группировки данных и анализа групп.
Освоить методы построения сводных таблиц и научиться с их помощью исследовать зависимости в признаках.
Научиться работать с данными, представленными в виде нескольких таблиц, и познакомиться с методами их объединения.

Вспомним, с какими данными мы работаем

В этом модуле мы продолжим нашу работу с датасетом о продажах объектов недвижимости в Мельбурне и его пригородах. Информацию об исходных данных вы можете найти здесь.

В прошлом модуле мы совершили множество преобразований над нашей таблицей — давайте вспомним их:

удалили столбцы index и Coordinates;
создали признак средней площади одной комнаты MeanRoomsSquare и ввели коэффициент соотношения площади здания к площади участка — AreaRatio;
преобразовали признак даты продажи Date в формат datetime и создали на его основе следующие столбцы: номер месяц продажи (MonthSale), номер дня недели продажи (WeekdaySale), «признак-мигалку» выходного дня (Weekend);
заменили признак года постройки здания YearBuilt на его возраст AgeBuilding;

извлекли из признака адреса объекта Address новый признак подтипа улицы StreetType и удалили столбец с адресом;
уменьшили число уникальных наименований агентств по недвижимости (SellerG), а также число пригородов (Suburb);
выделили категориальные признаки и преобразовали их в тип данных сategory;
заменили сокращённые названия категорий признака типа объекта Type на их полные названия (h — house, t —  townhouse, u — unit).

Скачать преобразованный набор данных в формате csv можно здесь (csv-файл находится в zip-архиве — распакуйте архив, прежде чем продолжать работу!)

Прочитаем датасет и выведем первые пять его строк:

In [1]:
import pandas as pd

melb_df = pd.read_csv('data/melb_data_fe.csv')
melb_df.head()

Suburb  Rooms   Type      Price Method SellerG        Date  Distance  \
0  Abbotsford      2  house  1480000.0      S  Biggin  2016-03-12       2.5   
1  Abbotsford      2  house  1035000.0      S  Biggin  2016-04-02       2.5   
2  Abbotsford      3  house  1465000.0     SP  Biggin  2017-04-03       2.5   
3  Abbotsford      3  house   850000.0     PI  Biggin  2017-04-03       2.5   
4  Abbotsford      4  house  1600000.0     VB  Nelson  2016-04-06       2.5   

   Postcode  Bedroom  ...  Longtitude             Regionname  Propertycount  \
0      3067        2  ...    144.9984  Northern Metropolitan           4019   
1      3067        2  ...    144.9934  Northern Metropolitan           4019   
2      3067        3  ...    144.9944  Northern Metropolitan           4019   
3      3067        3  ...    144.9969  Northern Metropolitan           4019   
4      3067        3  ...    144.9941  Northern Metropolitan           4019   

   MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding WeekdaySale  StreetType  \
0            25.20 -0.231707          3           46           5          St   
1            15.80 -0.327660          4          116           5          St   
2            18.75  0.056338          4          117           0          St   
3            15.75  0.145455          4           47           0       other   
4            17.75  0.083969          4            2           2          St   

   Weekend  
0        1  
1        1  
2        0  
3        0  
4        0  

[5 rows x 26 columns]

Файл подгрузился верно, однако есть один нюанс. Давайте посмотрим на информацию о столбцах с помощью метода info():



In [2]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Suburb           13580 non-null  object 
 1   Rooms            13580 non-null  int64  
 2   Type             13580 non-null  object 
 3   Price            13580 non-null  float64
 4   Method           13580 non-null  object 
 5   SellerG          13580 non-null  object 
 6   Date             13580 non-null  object 
 7   Distance         13580 non-null  float64
 8   Postcode         13580 non-null  int64  
 9   Bedroom          13580 non-null  int64  
 10  Bathroom         13580 non-null  int64  
 11  Car              13580 non-null  int64  
 12  Landsize         13580 non-null  float64
 13  BuildingArea     13580 non-null  float64
 14  CouncilArea      12211 non-null  object 
 15  Lattitude        13580 non-null  float64
 16  Longtitude       13580 non-null  float64
 17  Regionname  

None

Обратите внимание, что наше преобразование столбцов к типам datetime и category «слетело».

Почему это произошло?

Ответ на самом деле очень прост: csv-файл не хранит в себе информацию о типах данных столбцов, поэтому при чтении Pandas автоматически определяет тип данных столбца. Не забывайте об этом, обмениваясь преобразованными данными с вашими коллегами.

А пока предлагаем вам самим совершить преобразование столбцов к нужным типам данных ↓

### Задание 1.1
1 point possible (graded)
Преобразуйте столбец Date в формат datetime и выделите квартал (quarter) продажи объектов недвижимости. Найдите второй по популярности квартал продажи. В качестве ответа запишите число объектов, проданных в этом квартале.

Ответ:
4359

In [6]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'])
display(melb_df['Date'])

quarter_sale = melb_df['Date'].dt.quarter
display(quarter_sale.value_counts())

0       2016-03-12
1       2016-04-02
2       2017-04-03
3       2017-04-03
4       2016-04-06
           ...    
13575   2017-08-26
13576   2017-08-26
13577   2017-08-26
13578   2017-08-26
13579   2017-08-26
Name: Date, Length: 13580, dtype: datetime64[ns]

Date
3    4873
2    4359
4    2329
1    2019
Name: count, dtype: int64

### Задание 1.2
1 point possible (graded)
Преобразуйте все столбцы, в которых меньше 150 уникальных значений, в тип данных category, исключив из преобразования столбцы Date, Rooms, Bedroom, Bathroom, Car.
В качестве ответа запишите результирующее количество столбцов, которые имеют тип данных category.

Ответ:
10

In [8]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car'] # список столбцов, которые мы не берём во внимание
max_unique_count = 150 # задаём максимальное число уникальных категорий
for col in melb_df.columns: # цикл по именам столбцов
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude: # проверяем условие
        melb_df[col] = melb_df[col].astype('category') # преобразуем тип столбца
display(melb_df.info())

# Выбор столбцов с типом данных 'category' 
category_columns = melb_df.select_dtypes(include=['category']) 
# Подсчет количества столбцов с типом данных 'category' 
display(len(category_columns.columns))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  category      
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  category      
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

None

10

## 2. Сортировка данных в DataFrame


✍ Часто бывает так, что таблицу или промежуточный результат операций с ней необходимо отсортировать по какому-то критерию. Например, для отчётности вам необходимо предоставить список проданных объектов недвижимости, отсортированный по возрастанию цены или дате продажи. Рассмотрим основные подходы к решению таких задач.

### Метод sort_values()

Для сортировки значений в DataFrame по значениям одного или нескольких столбцов используется метод sort_values().

Основные параметры метода sort_values()

by — имя или список имён столбцов, по значениям которых производится сортировка.

axis — ось, по которой производится сортировка (0 — строки, 1 — столбцы). По умолчанию сортировка производится по строкам.

ascending — сортировка по возрастанию (от меньшего к большему). По умолчанию параметр выставлен на True, для сортировки по убыванию (от большего к меньшему) необходимо выставить его на False.

ignore_index — создаются ли новые индексы в таблице. По умолчанию выставлен на False и сохраняет индексы изначальной таблицы.

inplace — производится ли замена исходной таблицы на отсортированную. По умолчанию параметр выставлен на False, то есть замены не производится. Чтобы переопределить исходную таблицу на отсортированную, необходимо выставить этот параметр на True.


Сортировка по значениям одного столбца

Приведём несколько примеров сортировки нашей таблицы с недвижимостью.

Отсортируем таблицу по возрастанию цены объектов недвижимости (Price):

In [9]:
melb_df.sort_values(by='Price').head(10)


Suburb  Rooms   Type     Price Method        SellerG  \
2652        Footscray      1   unit   85000.0     PI        Burnham   
1805            other      4  house  131000.0     PI          other   
7303           Albion      1   unit  145000.0     PI         Biggin   
1927           Coburg      4  house  145000.0     PI         Jellis   
7940         Hawthorn      1   unit  160000.0     VB            HAR   
12666       Brunswick      1   unit  170000.0     VB         Nelson   
8811        Footscray      1   unit  170000.0     PI        Burnham   
7293           Albion      1   unit  185000.0      S  hockingstuart   
8504   West Footscray      1   unit  185000.0     PI            Jas   
7305           Albion      2   unit  190000.0     SP        Burnham   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
2652  2016-03-09       6.4      3011        1  ...   144.89000   
1805  2017-02-25       8.9      3162        4  ...   145.02420   
7303  2016-05-28      13.9      3020        2  ...   144.82660   
1927  2016-04-06       7.8      3058        3  ...   144.96580   
7940  2017-08-04       4.6      3122        1  ...   145.03730   
12666 2017-09-16       5.2      3056        1  ...   144.95188   
8811  2017-01-07       5.1      3011        1  ...   144.89587   
7293  2016-08-22      13.9      3020        1  ...   144.82720   
8504  2017-04-29       8.2      3012        1  ...   144.86720   
7305  2016-07-30      13.9      3020        2  ...   144.82390   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
2652    Western Metropolitan           7570        42.000000  1.000000   
1805   Southern Metropolitan           2379        17.222222 -0.525994   
7303    Western Metropolitan           2185        31.500000  0.555556   
1927   Northern Metropolitan          11204        20.500000 -0.531429   
7940   Southern Metropolitan          11308        42.000000 -0.437500   
12666  Northern Metropolitan          11918        42.000000 -0.816860   
8811    Western Metropolitan           7570         8.666667 -0.071429   
7293    Western Metropolitan           2185        14.333333 -0.964017   
8504    Western Metropolitan           5058        42.000000  1.000000   
7305    Western Metropolitan           2185        25.200000  1.000000   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
2652           3            9           2          St        0  
1805           2           97           5          St        1  
7303           5           46           5          St        1  
1927           4          106           2          Rd        0  
7940           8            8           4          St        0  
12666          9           47           5          St        1  
8811           1            4           5          St        1  
7293           8           41           0          Rd        0  
8504           4           47           5          St        1  
7305           7           46           5          St        1  

[10 rows x 26 columns]

Мы вывели десять строк таблицы, чтобы убедиться в верном порядке сортировки. Также обратите внимание на индексы таблицы — их значения сохранились из исходной таблицы.

А теперь отсортируем таблицу по убыванию (от самой последней до самой первой) даты продажи объекта (Date). Для этого выставим параметр ascending на False:

In [10]:
melb_df.sort_values(by='Date', ascending=False)

Suburb  Rooms   Type      Price Method    SellerG       Date  \
11144     Northcote      4  house  1955000.0     SP    McGrath 2017-12-08   
11217  Surrey Hills      3  house  1775000.0     PI     Jellis 2017-12-08   
11206      St Kilda      4  house  1600000.0     VB       Gary 2017-12-08   
11207    Strathmore      7  house  2000000.0     SP     Nelson 2017-12-08   
11208    Strathmore      4  house  1610000.0      S      other 2017-12-08   
...             ...    ...    ...        ...    ...        ...        ...   
1555     Camberwell      4  house  2650000.0      S     Jellis 2016-03-09   
4510       Oak Park      3  house   892000.0      S     Nelson 2016-03-09   
4388      Northcote      3  house  1200000.0      S     Nelson 2016-03-09   
6184   Surrey Hills      3  house  1205000.0      S  Fletchers 2016-01-28   
6183   Surrey Hills      2   unit   813000.0      S  Fletchers 2016-01-28   

       Distance  Postcode  Bedroom  ...  Longtitude             Regionname  \
11144       5.3      3070        4  ...   144.99375  Northern Metropolitan   
11217      10.2      3127        3  ...   145.08968  Southern Metropolitan   
11206       5.0      3182        4  ...   144.98324  Southern Metropolitan   
11207       8.2      3041        7  ...   144.90587   Western Metropolitan   
11208       8.2      3041        4  ...   144.92243   Western Metropolitan   
...         ...       ...      ...  ...         ...                    ...   
1555        7.8      3124        4  ...   145.06860  Southern Metropolitan   
4510       11.5      3046        3  ...   144.91740  Northern Metropolitan   
4388        5.5      3070        3  ...   144.99090  Northern Metropolitan   
6184       11.2      3127        3  ...   145.10060  Southern Metropolitan   
6183       11.2      3127        2  ...   145.10230  Southern Metropolitan   

       Propertycount  MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding  \
11144          11364        12.600000  0.923664         12           47   
11217           5457        15.750000 -0.721854         12           47   
11206          13240        16.000000 -0.586028         12          107   
11207           3284        20.882353 -0.262721         12           18   
11208           3284        24.909091 -0.517181         12           62   
...              ...              ...       ...        ...          ...   
1555            8920        18.900000 -0.550535          3          126   
4510            2651        18.000000 -0.714609          3           46   
4388           11364        17.857143 -0.264706          3           96   
6184            5457        18.000000 -0.590909          1           46   
6183            5457        25.200000  0.076923          1           46   

      WeekdaySale  StreetType  Weekend  
11144           4          St        0  
11217           4          Rd        0  
11206           4          St        0  
11207           4          Rd        0  
11208           4          St        0  
...           ...         ...      ...  
1555            2          Rd        0  
4510            2          St        0  
4388            2          St        0  
6184            3          Rd        0  
6183            3          Rd        0  

[13580 rows x 26 columns]

Сортировка по значениям нескольких столбцов

Для сортировки по значениям нескольких столбцов необходимо передать названия этих столбцов в параметр by в виде списка. При этом важно обращать внимание на порядок следования столбцов.

Так, например, отсортируем таблицу сначала по возрастанию расстояния от центра города (Distance), а затем — по возрастанию цены объекта (Price). Для того чтобы вывод был более наглядным, выделим каждую десятую строку из столбцов Distance и Price результирующей таблицы:

In [11]:
melb_df.sort_values(by=['Distance', 'Price']).loc[::10, ['Distance', 'Price']]

Distance      Price
11428       0.0   387000.0
10512       0.7   600000.0
5727        1.2   485000.0
8671        1.2   595000.0
5736        1.2   740000.0
...         ...        ...
12011      38.0   680000.0
10673      38.0   810000.0
13429      38.0  1155000.0
11102      41.0   650000.0
12728      45.9   650000.0

[1358 rows x 2 columns]

Мы получили таблицу, отсортированную по возрастанию расстояния до центра города. Если встречаются объекты недвижимости, у которых расстояние оказывается одинаковым, то внутри такой группы производится сортировка по цене объекта.

Ради интереса попробуйте поменять порядок следования столбцов в параметре by метода sort_values() и сравните результат. 

### Комбинирование сортировки с фильтрацией

А теперь рассмотрим применение сортировки на практике.

Предположим, компания McGrath поручила нам восстановить хронологию продаж таунхаусов, у которых площадь участка существенно больше площади здания, чтобы понять, как часто компания справляется с таким сложным видом объектов. Объекты, проданные в один и тот же день, мы бы хотели сортировать по значению коэффициента соотношения площадей. 

Найдём информацию о таунхаусах (Type), проданных компанией (SellerG) McGrath, у которых коэффициент соотношения площадей здания и участка (AreaRatio) меньше -0.8. Результат отсортируем по дате продажи (Date) в порядке возрастания, а после проведём сортировку по убыванию коэффициента соотношения площадей. Также обновим старые индексы на новые, установив параметр ignore_index на True. Для наглядности результата выберем из таблицы только столбцы Data и AreaRatio:

In [30]:
mask1 = melb_df['AreaRatio'] < -0.8
mask2 = melb_df['Type'] == 'townhouse'
mask3 = melb_df['SellerG'] == 'McGrath'
melb_df[mask1 & mask2 & mask3].sort_values(
    by=['Date', 'AreaRatio'],
    ascending=[True, False],
    ignore_index=True
).loc[:, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

Итак, мы рассмотрели основные методы сортировки данных. Сама по себе сортировка используется не так часто, однако в комбинации с другими методами обработки данных она является ценным инструментом. 

Предлагаем вам попрактиковаться в сортировке данных, выполнив несколько заданий ↓

Примечание. Старайтесь не сочетать фильтрацию и метод sort_values() с параметром inplace=True, так как в таком случае у вас возникнет предупреждение (warning) SettingWithCopyWarning: melb_df[melb_df['Rooms'] > 5].sort_values(inplace=True, by=['Rooms']):

C:\Users\Andrey\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingwithCopyWarning:

A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/ indexing. html#retu rning-a-view-versus-a-copy

    return func(*args, **kwargs)
Стоит отметить, что это не ошибка и код в таком случае отработает. Однако Pandas предупреждает вас о том, что при использовании такого кода дальнейшие результаты могут быть неожиданными.

Это предупреждение предназначено для обозначения операций «цепного присваивания». Это ситуация, когда вы пытаетесь напрямую изменить подмножество исходных данных. В нашем случае мы пытаемся отсортировать данные с заменой (об этом говорит параметр inplace=True) исходной таблицы на отсортированную.

Чтобы не возникало подобных конфликтов, необходимо использовать метод copy() для явного создания копии отфильтрованного подмножества исходных данных и работать уже с ней (производить сортировку):

filtered = melb_df[melb_df['Rooms'] > 5].copy()

filtered.sort_values(inplace=True, by=['Rooms'])

filtered.head()
Почитать о предупреждении SettingWithCopyWarning можно здесь.

Задание 2.2
1 point possible (graded)
Произведите сортировку столбца AreaRatio по возрастанию. При этом индексы полученной таблицы замените на новые. Какое значение площади здания находится в строке 1558? Ответ округлите до целого числа.

Ответ:
126


In [16]:
melb_df.sort_values(by=['AreaRatio'], ascending=[True], ignore_index=False).loc[1558, ['BuildingArea']]

BuildingArea    126.0
Name: 1558, dtype: object

### Задание 2.3
1 point possible (graded)
Найдите таунхаусы (Type) с количеством жилых комнат (Rooms) больше 2. Отсортируйте полученную таблицу сначала по возрастанию числа комнат, а затем по убыванию средней площади комнат (MeanRoomsSquare). Индексы таблицы замените на новые. Какая цена будет у объекта в строке 18? Ответ запишите в виде целого числа.

Ответ:
1300000

In [14]:
melb_df['Type'].value_counts()

Type
house        9449
unit         3017
townhouse    1114
Name: count, dtype: int64

In [ ]:
mask1 = melb_df['Type'] == 'townhouse'
mask2 = melb_df['Rooms'] > 2

display(melb_df[mask1 & mask2].sort_values(by=['Rooms', 'MeanRoomsSquare'], ascending=[True, False], ignore_index=True).loc[18:18, ['Price']])



Suburb  Rooms       Type      Price Method   SellerG  \
21         Abbotsford      3  townhouse   900000.0      S     other   
38       Airport West      3  townhouse   840000.0     PI    Nelson   
42       Airport West      3  townhouse   700000.0      S      Brad   
56       Airport West      3  townhouse   700000.0     PI      Brad   
67       Airport West      3  townhouse   540000.0     SP    Nelson   
...               ...    ...        ...        ...    ...       ...   
13437           other      4  townhouse   655000.0      S    Nelson   
13485           other      4  townhouse  1655000.0      S    Buxton   
13504  Mount Waverley      3  townhouse  1035000.0     SP   McGrath   
13521  Port Melbourne      4  townhouse  2455000.0     SP  Marshall   
13526       Reservoir      3  townhouse   640000.0     PI       Ray   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
21    2017-03-18       2.5      3067        3  ...   144.99590   
38    2016-03-09      13.5      3042        3  ...   144.87800   
42    2017-04-03      13.5      3042        3  ...   144.88370   
56    2016-10-15      13.5      3042        3  ...   144.88700   
67    2016-06-18      13.5      3042        2  ...   144.88020   
...          ...       ...       ...      ...  ...         ...   
13437 2017-08-26      12.9      3043        4  ...   144.90187   
13485 2017-08-26      11.4      3204        4  ...   145.04865   
13504 2017-08-26      14.2      3149        3  ...   145.12748   
13521 2017-08-26       3.5      3207        4  ...   144.94840   
13526 2017-08-26      12.0      3073        3  ...   145.01967   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
21     Northern Metropolitan           4019        15.750000  1.000000   
38      Western Metropolitan           3464        28.125000 -0.147727   
42      Western Metropolitan           3464        16.750000 -0.281501   
56      Western Metropolitan           3464        16.625000 -0.107383   
67      Western Metropolitan           3464        14.428571 -0.229008   
...                      ...            ...              ...       ...   
13437   Western Metropolitan           1071        19.545455  0.116883   
13485  Southern Metropolitan           2397        12.600000 -0.482546   
13504   Eastern Metropolitan          13366        17.250000 -0.262032   
13521  Southern Metropolitan           8648         0.000000 -1.000000   
13526  Northern Metropolitan          21650        18.285714 -0.272727   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
21             3            7           5          St        1  
38             3            0           2          Rd        0  
42             4            8           0          St        0  
56            10            7           5          St        1  
67             6            6           5          Rd        1  
...          ...          ...         ...         ...      ...  
13437          8            0           5          Dr        1  
13485          8           47           5          Av        1  
13504          8           11           5          Rd        1  
13521          8            7           5          St        1  
13526          8            9           5          St        1  

[784 rows x 26 columns]

Price
18  1300000.0

##  3. Группировка данных в DataFrame

✍ Одна из основных задач анализа данных — это группировка данных и сравнение показателей в группах. Например, нам необходимо сравнить средний уровень заработной платы в зависимости от пола/уровня образования. Или же мы хотим проследить, какая группа клиентов приносит нам наибольший доход, чтобы направить своё внимание на эту группу.

В некоторых случаях группировки может быть достаточно, чтобы ответить на вопросы бизнеса. В других случаях это может стать первым шагом в более сложном анализе. Так, например, благодаря группировке мы можем выявлять признаки, которые не несут статистической значимости, или признаки, которые вносят наибольший вклад. 

Так или иначе, владение группировкой — важный навык, который открывает новые возможности по работе с данными.

### Метод groupby()

В библиотеке Pandas для группировки данных по одному или нескольким признакам можно использовать метод groupby().

Кликните на плашку, чтобы увидеть информацию ↓

Основные параметры метода groupby()

by — имя или список имён столбцов, по которым производится группировка.

axis — ось, по которой производится группировка (0 — строки, 1 — столбцы). По умолчанию группировка производится по строкам.

as_index — добавляется ли дополнительный индекс к таблице. По умолчанию установлен на True.

Метод groupby() возвращает объект DataFrameGroupBy, который хранит в себе информацию о том, какие строки относятся к определённой группе, и сам по себе не представляет для нас интереса. 

Однако к этому объекту можно применять уже знакомые нам агрегирующие методы (mean, median, sum и т. д.), чтобы рассчитывать показатели внутри каждой группы.

Посмотрим на общую схему работы метода groupby() в Pandas:

Сначала мы разделяем данные на группы с помощью метода groupby(), после чего к каждой группе применяем агрегацию и объединяем результаты в новую таблицу.

Группировка данных по одному критерию с одной агрегацией

Рассмотрим группировку данных на примере нашей таблицы с недвижимостью.

Применим агрегирующую функцию среднего к результату работы groupby(). В качестве столбца для группировки возьмём столбец типа объекта недвижимости (Type):

In [34]:
melb_df.groupby(by='Type').mean(numeric_only=True)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\2774055988.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby(by='Type').mean(numeric_only=True)


Rooms         Price   Distance     Postcode   Bedroom  Bathroom  \
Type                                                                            
house      3.260874  1.242665e+06  10.979479  3104.080643  3.229336  1.613822   
townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183  1.809695   
unit       1.963871  6.051275e+05   7.607391  3110.797481  1.966523  1.183295   

                Car    Landsize  BuildingArea  Lattitude  Longtitude  \
Type                                                                   
house      1.772674  617.181924    152.162553 -37.803795  144.994700   
townhouse  1.555655  279.606822    134.649710 -37.815782  144.996489   
unit       1.128936  477.314219    102.235863 -37.823710  144.996363   

           Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
Type                                                               
house        7259.025505        18.996731  -0.490031    55.669700  
townhouse    7094.459605        18.569847  -0.094916    26.690305  
unit         8199.280080        21.068242   0.319883    39.703016

Мы получили таблицу, на пересечении строк и столбцов которой находятся средние значения каждого числового признака в наших данных.

Обратите внимание на структуру получившейся таблицы: теперь на месте индексов стоят значения типа объекта недвижимости Type (house, townhouse, unit).

Примечание. Если мы хотим видеть тип объекта в качестве отдельного столбца таблицы, мы можем выставить параметр as_index на False:

Img

Как правило, нам не нужна информация обо всех столбцах, поэтому агрегирующие методы можно применять только к интересующему нас столбцу. Например, давайте сравним средние цены на объекты в зависимости от их типа:

In [35]:
melb_df.groupby('Type')['Price'].mean()

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\2328732824.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby('Type')['Price'].mean()


Type
house        1.242665e+06
townhouse    9.337351e+05
unit         6.051275e+05
Name: Price, dtype: float64

Примечание. Обратите внимание, что, так как мы считаем только один показатель (среднее) для одного столбца, в результате мы получаем объект Series.

Из этой маленькой таблицы видно, что наибольшей средней ценой обладают объекты типа house (дома, коттеджи, виллы). Следовательно, можно сделать вывод, что тип постройки является значимым фактором при определении цены объекта недвижимости.

Теперь давайте выясним, какие регионы (Regionname) наиболее удалены от центра Мельбурна.

Для этого найдём минимальное значение расстояния от центра города до объекта в зависимости от его региона. Результат отсортируем по убыванию расстояния:

In [36]:
melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\3783769093.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)


Regionname
Western Victoria              29.8
Eastern Victoria              25.2
Northern Victoria             21.8
South-Eastern Metropolitan    14.7
Eastern Metropolitan           7.8
Western Metropolitan           4.3
Southern Metropolitan          0.7
Northern Metropolitan          0.0
Name: Distance, dtype: float64

Итак, наиболее удалёнными являются все регионы Victoria.

### Группировка данных по одному критерию с несколькими агрегациями

Чтобы рассчитать несколько агрегирующих методов, можно воспользоваться методом agg(), который принимает список строк с названиями агрегаций.

Давайте построим таблицу для анализа продаж по месяцам. Для этого найдём количество продаж, а также среднее и максимальное значения цен объектов недвижимости (Price), сгруппированных по номеру месяца продажи (MonthSale). Результат отсортируем по количеству продаж в порядке убывания:

In [37]:
melb_df.groupby('MonthSale')['Price'].agg(
    ['count', 'mean', 'max']
).sort_values(by='count', ascending=False)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\3700124206.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby('MonthSale')['Price'].agg(


count          mean        max
MonthSale                                
8           1850  1.056371e+06  6500000.0
7           1835  9.314698e+05  9000000.0
5           1644  1.097807e+06  8000000.0
6           1469  1.068981e+06  7650000.0
3           1408  1.146762e+06  5600000.0
4           1246  1.050479e+06  5500000.0
9           1188  1.126349e+06  6400000.0
10           854  1.135970e+06  6250000.0
11           750  1.142503e+06  5050000.0
12           725  1.144737e+06  5700000.0
2            333  1.169051e+06  4735000.0
1            278  9.397921e+05  5200000.0

Примечание. Обратите внимание, что, так как мы считаем несколько показателей для одного столбца, в результате мы получаем объект DataFrame.

В результате применения метода agg(), в который мы передали список с названиями интересующих нас агрегирующих функций, мы получаем DataFrame со столбцами count, mean и max, где для каждого месяца рассчитаны соответствующие параметры. Результат сортируем по столбцу count.

?
Какие интересные выводы можно сделать из этой таблицы?

1
Пик продаж приходится на период весна-лето.

2
Средняя цена продаваемых объектов относительно стабильна и находится в пределах 1 млн. австралийских долларов с небольшими отклонениями (около 100 тыс. влево и вправо).

3
Прослеживается некоторая зависимость между сезоном и максимальной ценой объектов: в месяцы с большим спросом на объекты недвижимости цена также имеет наибольшие показатели. Можно сделать предположение, что это связано с повышением цен на элитные дома в периоды большого спроса.

Примечание. Если вам нужна полная информация обо всех основных статистических характеристиках внутри каждой группы, вы можете воспользоваться методом agg(), передав в качестве его параметра строку 'describe':

melb_df.groupby('MonthSale')['Price'].agg('describe')

In [38]:
melb_df.groupby('MonthSale')['Price'].agg('describe')

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\701566850.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby('MonthSale')['Price'].agg('describe')


count          mean            std       min       25%        50%  \
MonthSale                                                                       
1           278.0  9.397921e+05  577668.924214  170000.0  570500.0   795000.0   
2           333.0  1.169051e+06  671564.357417  131000.0  710000.0  1020000.0   
3          1408.0  1.146762e+06  709573.596867   85000.0  680000.0   945000.0   
4          1246.0  1.050479e+06  591892.902979  145000.0  655000.0   905500.0   
5          1644.0  1.097807e+06  668492.867996  145000.0  650000.0   905000.0   
6          1469.0  1.068981e+06  606010.069052  222000.0  660000.0   900000.0   
7          1835.0  9.314698e+05  537390.803161  190000.0  586750.0   800000.0   
8          1850.0  1.056371e+06  619617.476541  160000.0  635000.0   892000.0   
9          1188.0  1.126349e+06  608734.690742  170000.0  725000.0   980000.0   
10          854.0  1.135970e+06  692950.251627  250000.0  652625.0   950000.0   
11          750.0  1.142503e+06  703666.227892  240000.0  671750.0   950000.0   
12          725.0  1.144737e+06  706549.007161  250000.0  667500.0   938000.0   

                 75%        max  
MonthSale                        
1          1111250.0  5200000.0  
2          1478000.0  4735000.0  
3          1400000.0  5600000.0  
4          1298750.0  5500000.0  
5          1371250.0  8000000.0  
6          1325000.0  7650000.0  
7          1150000.0  9000000.0  
8          1310000.0  6500000.0  
9          1360000.0  6400000.0  
10         1416500.0  6250000.0  
11         1399125.0  5050000.0  
12         1412000.0  5700000.0

После базовых математических функций наиболее частым агрегированием является подсчёт числа уникальных значений. Так, например, мы можем вычислить число уникальных риелторских компаний в зависимости от региона, чтобы понять, в каких регионах конкуренция на рынке недвижимости меньше. Это можно сделать, передав в параметр метода agg() строку 'nunique'. 

Более того, метод agg() поддерживает использование и других функций. Передадим дополнительно встроенную функцию set, чтобы получить множество из агентств недвижимости, которые работают в каждом из регионов:

In [40]:
melb_df.groupby('Regionname')['SellerG'].agg(
    		['nunique', set]
)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1626580071.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby('Regionname')['SellerG'].agg(


nunique  \
Regionname                            
Eastern Metropolitan             26   
Eastern Victoria                 11   
Northern Metropolitan            40   
Northern Victoria                11   
South-Eastern Metropolitan       25   
Southern Metropolitan            38   
Western Metropolitan             34   
Western Victoria                  6   

                                                                          set  
Regionname                                                                     
Eastern Metropolitan        {Noel, Barry, Gary, Woodards, McGrath, Stockda...  
Eastern Victoria            {Ray, Barry, Eview, Harcourts, O'Brien, Fletch...  
Northern Metropolitan       {Barry, Woodards, Village, Miles, McGrath, Big...  
Northern Victoria           {Raine, Barry, Ray, Buckingham, HAR, LITTLE, Y...  
South-Eastern Metropolitan  {Noel, Thomson, Barry, Gary, Woodards, McGrath...  
Southern Metropolitan       {Noel, Thomson, Barry, Gary, Woodards, William...  
Western Metropolitan        {Barry, Woodards, Williams, Village, Stockdale...  
Western Victoria                 {Raine, Ray, HAR, YPA, hockingstuart, other}

Как и ожидалось, наименьшая конкуренция в наиболее удалённом регионе Western Victoria, а наибольшая — в центральном районе Northern Metropolitan.

✍ Методы группировки по одному признаку позволяют посмотреть на наши данные «в разрезе» группирующего признака и понять их взаимосвязь. Далее мы рассмотрим, как оценивать взаимосвязь между большим количеством признаков.

Предлагаем вам самим потренироваться в использовании группировки ↓

### Задание 3.1
1 point possible (graded)
Сгруппируйте данные по признаку количества комнат и найдите среднюю цену объектов недвижимости в каждой группе. В качестве ответа запишите количество комнат, для которых средняя цена наибольшая.

Ответ:
7

In [42]:
melb_df.groupby('Rooms')['Price'].mean().sort_values(ascending=True)

Rooms
1     4.338245e+05
2     7.750812e+05
10    9.000000e+05
3     1.076081e+06
4     1.445282e+06
8     1.602750e+06
6     1.849366e+06
5     1.870260e+06
7     1.920700e+06
Name: Price, dtype: float64

### Задание 3.2
1 point possible (graded)
Какой регион имеет наименьшее стандартное отклонение по географической широте (Lattitude)?
В качестве ответа запишите название этого региона.

Ответ:
Western Victoria

In [43]:
melb_df.groupby('Regionname')['Lattitude'].std().sort_values(ascending=True)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1140392200.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby('Regionname')['Lattitude'].std().sort_values(ascending=True)


Regionname
Western Victoria              0.011579
Southern Metropolitan         0.043080
Eastern Metropolitan          0.047890
Northern Metropolitan         0.049639
Western Metropolitan          0.051251
South-Eastern Metropolitan    0.073411
Northern Victoria             0.084455
Eastern Victoria              0.147067
Name: Lattitude, dtype: float64

### Задание 3.3
1 point possible (graded)
Какая риелторская компания (SellerG) имеет наименьшую общую выручку за период с 1 мая по 1 сентября (включительно) 2017 года?
Для ответа на этот вопрос рассчитайте сумму продаж (Price) каждой компании в заданный период.
Не забудьте перевести даты в формат datetime.

Ответ:
LITTLE             

In [58]:
start_date = '2017-05-01' 
end_date = '2017-09-01'

mask1 = melb_df['Date'] >= start_date
mask2 = melb_df['Date'] <= end_date

melb_df[mask1 & mask2].groupby('SellerG')['Price'].sum().sort_values(ascending=True).idxmin()



C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\2056989235.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df[mask1 & mask2].groupby('SellerG')['Price'].sum().sort_values(ascending=True).idxmin()


'LITTLE'

## 4. Сводные таблицы

 Сводные таблицы — это распространённый инструмент для агрегации данных.

→ Сводная таблица принимает на вход данные из отдельных столбцов и группирует их. В результате получается новая таблица, которая позволяет увидеть многомерное обобщение данных. Таким образом, благодаря сводным таблицам мы можем оценить зависимость между двумя и более признаками данных.

Мы чаще сталкиваемся со сводными таблицами, чем с обычными, в плоском виде, так как сводные таблицы удобнее для анализа и быстрых выводов, а также позволяют увидеть более общие зависимости между признаками, нежели простая группировка данных.

Инструмент сводных таблиц также широко популярен среди тех, кто использует Excel или какие-либо BI-системы.

### Метод groupby как способ построения сводных таблиц

На самом деле мы с вами уже строили простейшие одномерные сводные таблицы с помощью метода groupby — мы рассматривали сводную таблицу в контексте группировки по одному признаку. 

Например, мы уже умеем строить таблицу, которая показывает зависимость медианной цены и площади здания от числа комнат:

In [59]:
melb_df.groupby('Rooms')[['Price', 'BuildingArea']].median()

Price  BuildingArea
Rooms                         
1       385000.0         107.0
2       690000.0         126.0
3       950000.0         126.0
4      1285000.0         142.0
5      1660000.0         176.0
6      1800000.0         126.0
7      1496000.0         216.5
8      1515000.0         126.0
10      900000.0         126.0

Также можно построить таблицу, в которой мы будем учитывать не только число комнат, но и тип здания (Type). Для этого в параметрах метода groupby() укажем список из нескольких интересующих нас столбцов.

In [60]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean()

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1929913328.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby(['Rooms', 'Type'])['Price'].mean()


Rooms  Type     
1      house        8.668655e+05
       townhouse    5.927045e+05
       unit         3.899289e+05
2      house        1.017238e+06
       townhouse    7.101585e+05
       unit         6.104905e+05
3      house        1.109233e+06
       townhouse    9.847087e+05
       unit         8.505963e+05
4      house        1.462283e+06
       townhouse    1.217092e+06
       unit         1.037476e+06
5      house        1.877327e+06
       townhouse    1.035000e+06
       unit                  NaN
6      house        1.869508e+06
       townhouse             NaN
       unit         5.200000e+05
7      house        1.920700e+06
       townhouse             NaN
       unit                  NaN
8      house        1.510286e+06
       townhouse             NaN
       unit         2.250000e+06
10     house        9.000000e+05
       townhouse             NaN
       unit                  NaN
Name: Price, dtype: float64

В результате выполнения такого кода мы получаем Series, которая обладает несколькими уровнями индексов: первый уровень — число комнат, второй уровень — тип здания. Такая организация индексов называется иерархической. Вычисление параметра (средней цены) происходит во всех возможных комбинациях признаков.

Для того, чтобы финальный результат был представлен в виде сводной таблицы (первый группировочный признак по строкам, а второй — по столбцам), а не в виде Series с иерархическими индексами, к результату чаще всего применяют метод unstack(), который позволяет переопределить вложенный индекс в виде столбцов таблицы:

In [61]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean().unstack()

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\442157687.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  melb_df.groupby(['Rooms', 'Type'])['Price'].mean().unstack()


Type          house     townhouse          unit
Rooms                                          
1      8.668655e+05  5.927045e+05  3.899289e+05
2      1.017238e+06  7.101585e+05  6.104905e+05
3      1.109233e+06  9.847087e+05  8.505963e+05
4      1.462283e+06  1.217092e+06  1.037476e+06
5      1.877327e+06  1.035000e+06           NaN
6      1.869508e+06           NaN  5.200000e+05
7      1.920700e+06           NaN           NaN
8      1.510286e+06           NaN  2.250000e+06
10     9.000000e+05           NaN           NaN

В результате мы получаем сводную таблицу, столбцы в которой представляют типы домов (house, townhouse, unit), строки — число комнат, а на пересечении строк и столбцов находится средняя стоимость объекта с такими показателями.

?
Какие интересные выводы можно сделать из этой таблицы?

1
Пропуски в сводной таблице (NaN) говорят о том, что в наших данных нет соответствующих комбинаций признаков. Например, у нас нет информации о ценах на таунхаусы, где количество комнат больше пяти.

2
Наибольшей средней стоимостью (2,25 млн. австралийских долларов) обладают объекты типа unit с восемью жилыми комнатами. Наименьшая средняя стоимость — у однокомнатных домов типа unit (чуть меньше 400 тыс. австралийских долларов).

3
Сколько бы комнат ни было в доме, цена на объекты типа unit всегда ниже других (за исключением восьмикомнатных объектов).

### Метод pivot_table для построения сводных таблиц

На самом деле метод groupby редко используется при двух параметрах, так как для построения сводных таблиц существует специальный и более простой метод — pivot_table().

Кликните на плашку, чтобы увидеть информацию ↓

Основные параметры метода pivot_table()

values — имя столбца, по которому необходимо получить сводные данные, применяя агрегирующую функцию;

index — имя столбца, значения которого станут строками сводной таблицы;

columns — имя столбца, значения которого станут столбцами сводной таблицы;

aggfunc — имя или список имён агрегирующих функций (по умолчанию — подсчёт среднего, 'mean');

fill_value — значение, которым необходимо заполнить пропуски (по умолчанию пропуски не заполняются).

Давайте построим ту же самую таблицу, но уже с использованием метода pivot_table. В качестве параметра values укажем столбец Price, в качестве индексов сводной таблицы возьмём Rooms, а в качестве столбцов — Type. Агрегирующую функцию оставим по умолчанию (среднее). Дополнительно заменим пропуски в таблице на значение 0. Финальный результат для наглядности вывода округлим с помощью метода round() до целых.

In [62]:
melb_df.pivot_table(
    values='Price',
    index='Rooms',
    columns='Type',
    fill_value=0
).round()

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\3628189379.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  melb_df.pivot_table(


Type       house  townhouse       unit
Rooms                                 
1       866866.0   592705.0   389929.0
2      1017238.0   710158.0   610491.0
3      1109233.0   984709.0   850596.0
4      1462283.0  1217092.0  1037476.0
5      1877327.0  1035000.0        0.0
6      1869508.0        0.0   520000.0
7      1920700.0        0.0        0.0
8      1510286.0        0.0  2250000.0
10      900000.0        0.0        0.0

Несложно понять, что метод pivot_table() имеет преимущество перед группировкой по нескольким критериям. Оно заключается в наличии специальных аргументов для строк и столбцов сводной таблицы, благодаря чему уменьшается вероятность запутаться при построении более сложных (многомерных) сводных таблиц, о которых мы поговорим далее.

А теперь давайте проанализируем продажи в каждом из регионов в зависимости от того, будний был день или выходной. Для этого построим сводную таблицу, в которой строками будут являться названия регионов (Regionname), а в столбцах будет располагаться наш «признак-мигалка» выходного дня (Weekend), который равен 1, если день был выходным, и 0 — в противном случае. В качестве значений сводной таблицы возьмём количество продаж.

In [63]:
melb_df.pivot_table(
    values='Price',
    index='Regionname',
    columns='Weekend',
    aggfunc='count'
)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\2049665855.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  melb_df.pivot_table(


Weekend                        0     1
Regionname                            
Eastern Metropolitan         447  1024
Eastern Victoria              13    40
Northern Metropolitan       1258  2632
Northern Victoria             11    30
South-Eastern Metropolitan   123   327
Southern Metropolitan       1534  3161
Western Metropolitan         960  1988
Western Victoria               8    24

Из результирующей таблицы можно сделать два вывода:

1
Число продаж резко возрастает в выходные вне зависимости от региона (приблизительно в 2-3 раза). То есть вероятность того, что дом продадут в выходные, гораздо выше вероятности, что его продадут в будний день.

2
В отдалённых регионах (Victoria) коэффициент роста числа продаж выше, чем в центральных. Если в центральных регионах Metropolitan продажи по выходным в 2-2.5 раза выше, чем по будням, то в регионах Victoria число продаж в выходные вырастает примерно в 3 раза.

Такой рост можно даже попытаться объяснить логически: в выходные дни у людей появляется свободное время, чтобы доехать до отдалённых пригородов с целью покупки дома.

Разберём ещё один пример: найдём, как зависит средняя и медианная площадь участка (Landsize) от типа объекта (Type) и его региона (Regionname). Чтобы посмотреть несколько статистических параметров, нужно передать в аргумент aggfunc список из агрегирующих функций. Построим такую сводную таблицу, где пропущенные значения заменим на 0:

In [64]:
melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\186795090.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  melb_df.pivot_table(
C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\186795090.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  melb_df.pivot_table(


median                          mean              \
Type                        house townhouse   unit        house   townhouse   
Regionname                                                                    
Eastern Metropolitan        674.0     233.5  203.0   717.422847  269.440678   
Eastern Victoria            843.0       0.0  230.0  3108.960000    0.000000   
Northern Metropolitan       459.5     134.0    0.0   619.249092  317.325733   
Northern Victoria           724.0       0.0    0.0  3355.463415    0.000000   
South-Eastern Metropolitan  630.5     240.0  199.0   664.306701  212.160000   
Southern Metropolitan       586.0     246.0    0.0   569.643881  278.858824   
Western Metropolitan        531.0     198.0   62.0   507.883406  244.560669   
Western Victoria            599.5       0.0    0.0   655.500000    0.000000   

                                        
Type                              unit  
Regionname                              
Eastern Metropolitan        330.444444  
Eastern Victoria            295.333333  
Northern Metropolitan       495.026538  
Northern Victoria             0.000000  
South-Eastern Metropolitan  357.864865  
Southern Metropolitan       466.380245  
Western Metropolitan        557.637232  
Western Victoria              0.000000

Обратите внимание на добавление дополнительных индексов столбцов median и mean. Здесь медианное и среднее значения рассчитаны отдельно для каждой комбинации признаков.

Здесь в глаза бросаются объекты типа house в регионах Eastern Victoria и Northern Victoria — в них среднее и медиана отличаются более чем в три раза. Вероятно, это связано с тем, что в этих районах очень большой разброс цен: есть несколько объектов с гигантской площадью, а остальные объекты имеют небольшую площадь. Из-за этого среднее значение искажается, в то время как медиана нечувствительна к такому разбросу и не искажает результат.

### Многомерные сводные таблицы

До этого мы рассматривали, как некоторый статистический показатель может зависеть от двух признаков. Однако, как уже упоминалось, сводные таблицы позволяют наблюдать зависимость и от большего числа признаков. Такие сводные таблицы называются многомерными. 

Для того чтобы исследовать зависимость от большего числа признаков, можно передать список признаков в параметр index или параметр columns.

Давайте построим таблицу, в которой по индексам будут располагаться признаки метода продажи (Method) и типа объекта (Type), по столбцам — наименование региона (Regionname), а на пересечении строк и столбцов будет стоять медианная цена объекта (Price):

In [65]:
melb_df.pivot_table(
    values='Price',
    index=['Method','Type'],
    columns='Regionname',
    aggfunc='median',
    fill_value=0
)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1809242545.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  melb_df.pivot_table(


Regionname        Eastern Metropolitan  Eastern Victoria  \
Method Type                                                
PI     house                 1244000.0          780000.0   
       townhouse              760000.0               0.0   
       unit                   650000.0               0.0   
S      house                 1127000.0          675000.0   
       townhouse              828000.0               0.0   
       unit                   645750.0          492000.0   
SA     house                  932500.0          950000.0   
       townhouse              807500.0               0.0   
       unit                        0.0               0.0   
SP     house                 1050000.0          672500.0   
       townhouse              910000.0               0.0   
       unit                   515000.0          400000.0   
VB     house                 1100000.0          712500.0   
       townhouse              892500.0               0.0   
       unit                   500000.0               0.0   

Regionname        Northern Metropolitan  Northern Victoria  \
Method Type                                                  
PI     house                   900000.0           500000.0   
       townhouse               632500.0                0.0   
       unit                    410000.0                0.0   
S      house                   920000.0           555000.0   
       townhouse               750000.0                0.0   
       unit                    525500.0                0.0   
SA     house                   817500.0           540000.0   
       townhouse               425000.0                0.0   
       unit                    616000.0                0.0   
SP     house                   900000.0           521000.0   
       townhouse               690000.0                0.0   
       unit                    470000.0                0.0   
VB     house                  1050000.0           690000.0   
       townhouse               640000.0                0.0   
       unit                    450000.0                0.0   

Regionname        South-Eastern Metropolitan  Southern Metropolitan  \
Method Type                                                           
PI     house                        865000.0              1725000.0   
       townhouse                   1190000.0              1055000.0   
       unit                         525000.0               571250.0   
S      house                        883300.0              1611000.0   
       townhouse                    875000.0              1135000.0   
       unit                         606000.0               655000.0   
SA     house                        880000.0              1390000.0   
       townhouse                         0.0              1141000.0   
       unit                              0.0               580000.0   
SP     house                        770000.0              1521750.0   
       townhouse                    800000.0              1162500.0   
       unit                         601000.0               550000.0   
VB     house                        850000.0              1800000.0   
       townhouse                         0.0              1250000.0   
       unit                         700000.0               500000.0   

Regionname        Western Metropolitan  Western Victoria  
Method Type                                               
PI     house                  870000.0          630000.0  
       townhouse              670000.0               0.0  
       unit                   360000.0               0.0  
S      house                  870000.0          397500.0  
       townhouse              729000.0               0.0  
       unit                   489000.0               0.0  
SA     house                  772500.0               0.0  
       townhouse              467500.0               0.0  
       unit                   571000.0               0.0  
SP     house                  865000.0          360000.0  
       t

Первым индексом в таблице идёт метод продажи здания, далее для метода указывается тип недвижимости. По столбцам расположены регионы. В ячейках таблицы указана медианная цена для каждой такой комбинации.

Такие таблицы уже сложнее читать, однако с помощью них можно более глубоко исследовать закономерности. Например, можно видеть, что вне зависимости от метода продажи и региона цена на объекты типа house практически всегда выше, чем на объекты другого типа.

### Доступ к данным в сводной таблице

?
Как получить доступ к данным или произвести фильтрацию в сложной сводной таблице, где есть дополнительные индексы?

Давайте рассмотрим, что собой представляют столбцы сложной сводной таблицы.

Запишем сводную таблицу, которую мы создавали ранее в переменную pivot:

In [66]:
pivot = melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1672652905.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot = melb_df.pivot_table(
C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1672652905.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot = melb_df.pivot_table(


Выведем её столбцы с помощью атрибута columns:

In [67]:
pivot.columns

MultiIndex([('median',     'house'),
            ('median', 'townhouse'),
            ('median',      'unit'),
            (  'mean',     'house'),
            (  'mean', 'townhouse'),
            (  'mean',      'unit')],
           names=[None, 'Type'])

В результате мы получаем объект MultiIndex. Этот объект хранит в себе шесть комбинаций пар столбцов (два статистических параметра и три типа здания), то есть есть шесть возможных вариантов обращения к столбцам таблицы.

Мультииндексы раскрываются подобно вложенным словарям — по очереди, как матрёшка. Чтобы получить доступ к определённому столбцу, вы должны сначала обратиться к столбцу, который находится уровнем выше.

Так, из таблицы pivot мы можем получить средние значения площадей участков для типа здания unit, просто последовательно обратившись по имени столбцов:

In [68]:
display(pivot['mean']['unit'])

Regionname
Eastern Metropolitan          330.444444
Eastern Victoria              295.333333
Northern Metropolitan         495.026538
Northern Victoria               0.000000
South-Eastern Metropolitan    357.864865
Southern Metropolitan         466.380245
Western Metropolitan          557.637232
Western Victoria                0.000000
Name: unit, dtype: float64

Аналогично производится и фильтрация данных. Например, если нам нужны регионы, в которых средняя площадь здания для домов типа house меньше их медианной площади, то мы можем найти их следующим образом:

In [69]:
mask = pivot['mean']['house'] < pivot['median']['house']
filtered_pivot = pivot[mask]
display(filtered_pivot)

median                        mean              \
Type                   house townhouse  unit       house   townhouse   
Regionname                                                             
Southern Metropolitan  586.0     246.0   0.0  569.643881  278.858824   
Western Metropolitan   531.0     198.0  62.0  507.883406  244.560669   

                                   
Type                         unit  
Regionname                         
Southern Metropolitan  466.380245  
Western Metropolitan   557.637232

Чтобы получить индексы отфильтрованной таблицы, можно воспользоваться атрибутом index и обернуть результат в список:

In [70]:
print(list(filtered_pivot.index))
# ['Southern Metropolitan', 'Western Metropolitan']

['Southern Metropolitan', 'Western Metropolitan']


✍ Таким образом, сводные таблицы изначально кажутся сложной структурой, но на самом деле это обычные DataFrame со вложенными индексами строк или столбцов. 

Умение читать и анализировать сложные сводные таблицы — это важный навык, который помогает проводить углублённый анализ данных.

### Задание 4.2
1 point possible (graded)
Составьте сводную таблицу, которая показывает зависимость медианной площади (BuildingArea) здания от типа объекта недвижимости (Type) и количества жилых комнат в доме (Rooms). Для какой комбинации признаков площадь здания наибольшая?
В качестве ответа запишите эту комбинацию (тип здания, число комнат) через запятую, без пробелов.

Ответ:
house,7

In [ ]:
pivot_table = melb_df.pivot_table(
    values='BuildingArea',
    index='Type',
    columns='Rooms',
    aggfunc=['median'],
    fill_value=0
)

display(pivot_table)

max_value = pivot_table.max().max()

max_index = pivot_table.idxmax()[1]
max_column = pivot_table.unstack().idxmax() [1] 


print(max_index, int(max_column), float(max_value))


C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1033856352.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_table = melb_df.pivot_table(


median                                                        
Rooms         1      2      3      4      5      6      7      8      10
Type                                                                    
house      126.0  126.0  126.0  141.0  177.0  126.0  216.5  126.0  126.0
townhouse   88.0  114.0  126.0  159.5  152.0    0.0    0.0    0.0    0.0
unit        69.5  110.0  126.0  126.0    0.0  171.0    0.0  126.0    0.0

house 7 216.5


C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\1033856352.py:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  max_index = pivot_table.idxmax()[1]


### Задание 4.3
1 point possible (graded)
Составьте сводную таблицу, которая показывает зависимость медианной цены объекта недвижимости (Price) от риелторского агентства (SellerG) и типа здания (Type).
Во вновь созданной таблице найдите агентство, у которого медианная цена для зданий типа unit максимальна. В качестве ответа запишите название этого агентства.

Ответ:

Nick
900000.0

In [74]:
melb_df.pivot_table(
    values='Price',
    index='SellerG',
    columns='Type',
    aggfunc=['median'],
    fill_value=0
)

C:\Users\stepu\AppData\Local\Temp\ipykernel_42288\2505052610.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  melb_df.pivot_table(


median                     
Type               house  townhouse      unit
SellerG                                      
Alexkarbon     1036250.0   935000.0  560000.0
Barry           830000.0   727500.0  502000.0
Bells           692500.0   430500.0  457500.0
Biggin         1151750.0   675000.0  548000.0
Brad            875000.0   700000.0  465000.0
Buckingham      830000.0   820000.0  553000.0
Burnham         922500.0        0.0  305000.0
Buxton         1323750.0  1030000.0  670000.0
C21             836500.0   768750.0  625500.0
Cayzer         1505000.0  1450000.0  707500.0
Chisholm       1520000.0   950000.0  640000.0
Collins        1566500.0   930000.0  592000.0
Douglas         684000.0   500000.0  494500.0
Eview           760000.0   597500.0  602000.0
Fletchers      1390000.0  1238000.0  653000.0
Gary           1370000.0   993500.0  550000.0
Greg           1330000.0   900000.0  547500.0
HAR             695000.0        0.0  518000.0
Harcourts       740000.0   653750.0  500000.0
Hodges         1500000.0  1105000.0  605500.0
Jas             935000.0   699000.0  455000.0
Jellis         1505000.0  1052000.0  591750.0
Kay            2220000.0  1365000.0  695000.0
LITTLE          970000.0   800000.0  447000.0
Love            780000.0   846000.0  440000.0
Marshall       1975000.0  1408500.0  715000.0
McDonald       1242500.0        0.0  502500.0
McGrath        1250000.0   810000.0  542500.0
Miles          1090000.0   721500.0  555000.0
Moonee          806000.0   715000.0  622000.0
Nelson         1055000.0   730000.0  530000.0
Nick           2025000.0   780000.0  900000.0
Noel           1400500.0   990000.0  693250.0
O'Brien         737504.0   430000.0  618500.0
Philip         1035000.0   701000.0  636000.0
Purplebricks    810000.0   927500.0  582500.0
RT             1640000.0  1400000.0  630000.0
RW              870000.0   630500.0  625500.0
Raine           745000.0   608000.0  447500.0
Ray             900000.0   650000.0  497500.0
Rendina        1190000.0   750000.0  459000.0
Stockdale       715500.0   550000.0  410000.0
Sweeney         850000.0   728000.0  381000.0
Thomson        1177000.0   500000.0  495000.0
Village         973500.0   727500.0  420000.0
Williams       1161000.0   735000.0  577000.0
Woodards       1261000.0   865500.0  600000.0
YPA             596250.0   498500.0  443500.0
hockingstuart  1055000.0   935000.0  566000.0
other           901000.0   840500.0  535000.0